### PR TITLE
Allow api module to fail

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Right now, the module will not fail when `return_result` is called with `status=False`, since it looks for `status == 'False'` (string instead of boolean).

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
api

### CODE REVIEW

**Comment by felixfontein** [comment]
This took a bit longer, since there were some problems with the tests. These should be fixed now as well :)

**Comment by NikolayDachev** [comment]
will check today

**Comment by felixfontein** [comment]
(I rebased after merging #37 so the PR's branch contains that PR as well. I didn't make any changes to this PR's code.)

**Comment by felixfontein** [comment]
Ok, this should either be merged in a bit more than a hour, or it has to skip this community.routeros release. Otherwise we might miss Ansible 4.2.0.

File: plugins/modules/api.py
Comment (NikolayDachev) [inline plugins/modules/api.py:461]: From def errors we return True when we get TrapError, I do that with purpose. No easy way to do pre check via api for different tasks in ROS ! For this reason I ask for example in #32. 

for this approach (original code)
**Pros:** No need to handle **false positive** failers (we have such in ros)
**Cons:** Yes we will get fail as is excepted ! however for the  **false positive**  cases users must handle them via roles/playbooks 

For example: 
`    "msg": "['failure: already have such address']"
`
This happen when you try to add same ip address to interface .. first run will be successful, second run will fail !
Now I know this is my strange opinion, however this save overhand for such false positive cases like add ip address.

If we allow to FAIL in such case, the role / playbook must have SPECIFIC task to check if such address already exist end  etc .. However this is only for add ip what about all other basilion different tasks in ros. For each different task playook/role must have specific pre-check 

I hope you get my point way I do that in first place .. Over all I'm ok to make it fail (change must be done in  def errors() ) Probably additional module var to Fail or Not to Fail will be a good idea :)

    def errors(self, e):
        if e.__class__.__name__ == 'TrapError':
            self.result['message'].append("%s" % e)
            self.return_result(False, **True**)
        self.result['message'].append("%s" % e)
        self.return_result(False, False)

**NOTE: I'm really sorry for this late review :(**

Comment (NikolayDachev) [reply plugins/modules/api.py:461]: When I think more, probably my opinion is wrong and we should make it to fail.

Comment (NikolayDachev) [reply plugins/modules/api.py:461]: one more think: 

Is the correct approach to make it "Fail" -- I fully agree however this will make idempotent of the roles nightmare, the reason is because ros don't make a difference between foe example "wrong ip address mask" and "ip address already exist " (as example).
For this reason this module don't support check_mode 

May be I miss something and more research is needed ... 
I really don't know ...

Comment (NikolayDachev) [reply plugins/modules/api.py:461]:
```
[admin@chr7b] > ip address/add address=192.168.12.3 netmask=255.255.256.0 interface=ether2
invalid value for argument netmask
[admin@chr7b] > ip address/add address=192.168.12.3 netmask=255.255.255.0 interface=ether2  
[admin@chr7b] > ip address/add address=192.168.12.3 netmask=255.255.255.0 interface=ether2 
failure: already have such address
```


same is via api module if we make it to fail
**first run:** ansible-playbook playbooks/ros_ip_addr_add.yml --limit chr7b --ask-vault-pass -vvvv

```
ok: [chr7b] => {
    "msg": {
        "changed": true,
        "failed": false,
        "msg": [
            "added: .id= *E"
        ]
    }
}
```

**second run:** ansible-playbook playbooks/ros_ip_addr_add.yml --limit chr7b --ask-vault-pass -vvvv

```
    "changed": false,
    "invocation": {
        "module_args": {
            "add": "broadcast= comment= copy-from= disabled=no netmask=255.255.255.0 network=192.168.192.0 address=192.168.192.20 interface=ether2",
            "ca_path": null,
            "cmd": null,
            "hostname": "chr7b",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "path": "ip address",
            "port": null,
            "query": null,
            "remove": null,
            "ssl": "false",
            "tls": false,
            "update": null,
            "username": "admin",
            "validate_cert_hostname": false,
            "validate_certs": true
        }
    },
    "msg": "['failure: already have such address']"
}

```

**wrong mask**  ansible-playbook playbooks/ros_ip_addr_add.yml --limit chr7b --ask-vault-pass -vvvv -e "ros_ip_addr_netmask=255.255.25.256"

```
    "changed": false,
    "invocation": {
        "module_args": {
            "add": "broadcast= comment= copy-from= disabled=no netmask=255.255.25.256 network=192.168.192.0 address=192.168.192.20 interface=ether2",
            "ca_path": null,
            "cmd": null,
            "hostname": "chr7b",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "path": "ip address",
            "port": null,
            "query": null,
            "remove": null,
            "ssl": "false",
            "tls": false,
            "update": null,
            "username": "admin",
            "validate_cert_hostname": false,
            "validate_certs": true
        }
    },
    "msg": "['invalid value for argument netmask']"
}

```

Comment (felixfontein) [reply plugins/modules/api.py:461]: > Is the correct approach to make it "Fail" -- I fully agree however this will make idempotent of the roles nightmare, the reason is because ros don't make a difference between for example "wrong ip address mask" and "ip address already exist " .
> For this reason this module don't support check_mode
> 
> May be I miss something and more research is needed ...
> I really don't know ...

Well, that's always the problem with low-level modules like `ansible.builtin.command` and `community.routeros.command` - it's a non-trivial amount of work to make them do something idempotent. If a command executed on a host (for `ansible.builtin.command`) or a router (for `community.routeros.command`) fails, the module should fail, and the user should be able to decide how to react on the failure. Simplify making the module ignore errors by default is definitely not right, since the module has no idea whether an error was intended, acceptable, or not.

Ansible provides `failed_when` for tasks (https://docs.ansible.com/ansible/latest/user_guide/playbooks_error_handling.html#defining-failure) to allow role/playbook authors to define when a command really failed. This can be used to distinguish between "acceptable errors" and "unacceptable errors". In your above example, `failed_when` can be something like `result is success or (result is failed and 'failure: already have such address' in result.msg)`. That way, you can make the module invocation idempotent, and still report errors on wrong input.

In the long run, it would be great if there would be more advanced modules for such tasks, which know about what they run and how to handle it - see #33.

Comment (NikolayDachev) [reply plugins/modules/api.py:461]: I agree in future will check with librouteros if is possible / usefully to utilize https://wiki.mikrotik.com/wiki/Manual:API#category somehow

Comment (jplitza) [reply plugins/modules/api.py:461]: I propose to explicitly mention the case of "already have ..." errors that will now be reported in the changelog, since that could be considered a regression.

**Comment by NikolayDachev** [comment]
Will try to check the test tonight

File: plugins/modules/api.py
Comment (felixfontein) [reply plugins/modules/api.py:461]: The changelog fragment definitely needs to be extended. I'm wondering whether we should make this even more prominent, or whether this should even be a change for community.routeros 2.0.0.

**Review by NikolayDachev — APPROVED** [review]

**Comment by NikolayDachev** [comment]
we can merge

**Comment by NikolayDachev** [comment]
> Ok, this should either be merged in a bit more than a hour, or it has to skip this community.routeros release. Otherwise we might miss Ansible 4.2.0.

Since this is more like bug fix I gu

**Comment by NikolayDachev** [comment]
closed by mistake

**Comment by felixfontein** [comment]
@heuels @NikolayDachev @jplitza I've updated the changelog fragment. Please take a look! Also if you think that this shouldn't just show up in a patch release, but potentially a minor or even a major release, please say so! This will very likely break people's roles and playbooks, when they rely on errors not resulting in task failures. We have to decide how breaking this is (all collections stick to [semantic versioning](https://semver.org/)).

Basically our options are:
- Put this in a patch release (1.2.1). We should only do this if we think this is an important bugfix that - even though it breaks people's workflows - should get out fast.
- Put this in a minor release (1.3.0). This is only useful if someone installs the collection manually (i.e. outside of the Ansible package) and uses a version restriction like `< 1.2.0` to prevent new features.
- Put this in a major release (2.0.0). This indicates a breaking change. A consequence is that 2.0.0 will not be included in any Ansible 4.x.0 release, but only in Ansible 5.0.0 (to be released in mid November; see https://docs.ansible.com/ansible/devel/roadmap/COLLECTIONS_5.html).

What do you think?

**Comment by NikolayDachev** [comment]
> @heuels @NikolayDachev @jplitza I've updated the changelog fragment. Please take a look! Also if you think that this shouldn't just show up in a patch release, but potentially a minor or even a major release, please say so! This will very likely break people's roles and playbooks, when they rely on errors not resulting in task failures. We have to decide how breaking this is (all collections stick to [semantic versioning](https://semver.org/)).
> 
> Basically our options are:
> 
> * Put this in a patch release (1.2.1). We should only do this if we think this is an important bugfix that - even though it breaks people's workflows - should get out fast.
> * Put this in a minor release (1.3.0). This is only useful if someone installs the collection manually (i.e. outside of the Ansible package) and uses a version restriction like `< 1.2.0` to prevent new features.
> * Put this in a major release (2.0.0). This indicates a breaking change. A consequence is that 2.0.0 will not be included in any Ansible 4.x.0 release, but only in Ansible 5.0.0 (to be released in mid November; see https://docs.ansible.com/ansible/devel/roadmap/COLLECTIONS_5.html).
> 
> What do you think?

1.3.0 look ok will give a some time for playbook/roles fixes

**Comment by felixfontein** [comment]
1.3.0 will not mean that the release will happen later, it only changes the semantic of the release in terms of semantic versioning.

**Comment by jplitza** [comment]
This definitely isn't a patch level change. I'm even tempted to say this is worth a new major version, since (roughly) a third of the collection changes a significant portion of its API (namely the return value, if you will). November isn't so far away, and this change isn't really pressing (it's been this way like forever, hasn't it?)

**Comment by heuels** [comment]
For me it's also tempting to say it's a major-level release, but still I would stick with a minor-lever here.

**Comment by NikolayDachev** [comment]
> 1.3.0 will not mean that the release will happen later, it only changes the semantic of the release in terms of semantic versioning.

Yes I know, what I try to point here was -> " This is only useful if someone installs the collection manually (i.e. outside of the Ansible package) and uses a version restriction like < 1.2.0 to prevent new features."

Other question: Why we should care about "ansible base" release date (except something new is pushed from it which we should implement). If I remember correctly this was the idea of small collections .. to not wait for "ansible base" release .. ?

**Comment by felixfontein** [comment]
> Other question: Why we should care about "ansible base" release date (except something new is pushed from it which we should implement). If I remember correctly this was the idea of small collections .. to not wait for "ansible base" release .. ?

The ansible-core release date does not matter at all. What potentially matters is the Ansible release date. Ansible is the community distribution "with batteries included", which consists of ansible-core and a set of collections, including community.routeros. Creating a 2.0.0 release now means that the community.routeros version included in Ansible 4.x.0 will be frozen, except if we backport things and create further community.routeros 1.x.y releases. While this is possible (and not that much work), backporting features is more complicated, so the community.routeros version included in Ansible 4.x.0 will probably only get bugfixes, but no more new features.

Considering that we aren't publishing much features anyway, that's probably ok though.

**Comment by NikolayDachev** [comment]
Well 1.3.0 ..

**Comment by felixfontein** [comment]
I've relabelled the patch as minor_changes in 11280093b68a6593c772c0754c826ee6db394c0c. I'll try to discuss this with some people in the Ansible community next week to get some more feedback; right now I'm tending a bit to make this a major release.

**Comment by NikolayDachev** [comment]
> I've relabelled the patch as minor_changes in 11280093b68a6593c772c0754c826ee6db394c0c. I'll try to discuss this with some people in the Ansible community next week to get some more feedback; right now I'm tending a bit to make this a major release.

Any news?

**Comment by felixfontein** [comment]
From the feedback I got, this should really be a major release, since it is a breaking change in functionality. If that's fine with everyone, I'll adjust the changelog fragment accordingly.

**Comment by felixfontein** [comment]
@heuels @NikolayDachev is this ok for you? If yes, I'll create a `stable-1` branch, merge this, and bump the next expected version in `main` to 2.0.0. (Then we can also think whether there are other breaking changes we want to do for a 2.0.0 release. Not sure if there is anything, but if there is something, that would be the time :) )

**Comment by NikolayDachev** [comment]
I'm OK with it, thanks a lot for the effort

**Comment by heuels** [comment]
Fine with me as well! Thanks for doing this.

**Comment by felixfontein** [comment]
@NikolayDachev @jplitza @heuels thanks for your reviews and feedback!

**Pull request merged**
